File: GridView/RowDetailsInGridView/RowDetailsInGridViewVB/RowDetailsLayout.vb
Imports Telerik.WinControls.UI

Public Class RowDetailsLayout
    Inherits TableViewRowLayout
    Private systemWidth As Integer
    Private availableSize As SizeF

    Public Overrides Sub InvalidateRenderColumns()
        MyBase.InvalidateRenderColumns()
        systemWidth = 0
        For Each column As GridViewColumn In Me.RenderColumns
            If Not (TypeOf column Is GridViewDataColumn) Then
                systemWidth += column.Width + Owner.CellSpacing
            End If
        Next column
        If systemWidth > 0 Then
            systemWidth -= Owner.CellSpacing
        End If
    End Sub

    Public Overrides Function MeasureRow(ByVal availableSize As SizeF) As SizeF
        Dim desiredSize As SizeF = MyBase.MeasureRow(availableSize)
        Me.availableSize = availableSize
        Return desiredSize
    End Function

    Public Overrides Function ArrangeCell(ByVal clientRect As RectangleF, ByVal cell As GridCellElement) As RectangleF
        Dim grid As RowDetailsGrid = CType(Owner.GridViewElement.GridControl, RowDetailsGrid)
        If cell.ColumnInfo Is grid.DetailsColumn Then
            cell.StretchHorizontally = True
            cell.BypassLayoutPolicies = True
            If cell.RowElement.IsCurrent Then
                Return New RectangleF(0, Owner.RowHeight, availableSize.Width - systemWidth, grid.DetailsRowHeight - Owner.RowHeight)
            Else
                Return RectangleF.Empty
            End If
        End If

        If TypeOf cell Is GridDataCellElement Then
            clientRect.Height = Owner.RowHeight
        End If

        Return MyBase.ArrangeCell(clientRect, cell)
    End Function
End Class

File: GridView/RowDetailsInGridView/RowDetailsInGridView/RowDetailsLayout.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Telerik.WinControls.UI;

namespace RowDetailsInGridView
{
    public class RowDetailsLayout : TableViewRowLayout
    {
        int systemWidth;
        SizeF availableSize;

        public override void InvalidateRenderColumns()
        {
            base.InvalidateRenderColumns();
            systemWidth = 0;
            foreach (GridViewColumn column in this.RenderColumns)
            {
                if (!(column is GridViewDataColumn))
                {
                    systemWidth += column.Width + Owner.CellSpacing;
                }
            }
            if (systemWidth > 0)
            {
                systemWidth -= Owner.CellSpacing;
            }
        }

        public override SizeF MeasureRow(SizeF availableSize)
        {
            SizeF desiredSize = base.MeasureRow(availableSize);
            this.availableSize = availableSize;
            return desiredSize;
        }

        public override RectangleF ArrangeCell(RectangleF clientRect, GridCellElement cell)
        {
            RowDetailsGrid grid = (RowDetailsGrid)Owner.GridViewElement.GridControl;

            if (grid.DetailsColumn == null)
            {
                return base.ArrangeCell(clientRect, cell);
            }

            int width = GetColumnWidth(grid.DetailsColumn) + grid.TableElement.CellSpacing;

            if (cell.RowInfo.IsCurrent)
            {
                if (cell.ColumnInfo == grid.DetailsColumn)
                {
                    float visibleColumnsWidth = 0;

                    foreach (GridViewColumn col in this.RenderColumns)
                    {
                        if (col is GridViewDataColumn)
                        {
                            visibleColumnsWidth += col.Width + grid.TableElement.CellSpacing;
                        }
                    }

                    return new RectangleF(clientRect.X, Owner.RowHeight, visibleColumnsWidth - 1, grid.DetailsRowHeight - Owner.RowHeight);
                }
                else
                {
                    RectangleF rect = base.ArrangeCell(clientRect, cell);
                    rect.Height = Owner.RowHeight;
                    rect.X -= width;

                    if (this.LastDataColumn == cell.ColumnInfo)
                    {
                        rect.Width += width;
                    }

                    return rect;
                }
            }
            else if (cell.ColumnInfo == grid.DetailsColumn)
            {
                return RectangleF.Empty;
            }

            if (cell.ColumnInfo is GridViewDataColumn)
            {
                RectangleF cellRect = base.ArrangeCell(clientRect, cell);
                cellRect.Height = Owner.RowHeight;
                cellRect.X -= width;

                if (this.LastDataColumn == cell.ColumnInfo)
                {
                    cellRect.Width += width;
                }

                return cellRect;
            }

            return base.ArrangeCell(clientRect, cell);
        }
    }
}

